File: c++/Pathfinder/src/Astar.cpp
/*
 * Astar.cpp
 *
 *  Created on: Apr 14, 2016
 *      Author: saverio
 */

#include "../headers/Astar.h"
#include <algorithm>
#include <set>
#include <unordered_set>

Astar :: Astar () {}

Astar :: Astar (Mat* grid) {
	this->grid = grid;
}

Astar :: ~Astar () {}

double Astar :: heuristic (Node* node1, Node* node2) {

	double a = pow((node1->row() - node2->row()), 2);
	double b = pow((node1->col() - node2->col()), 2);

	double dist = sqrt(a + b);

	return dist;

}

bool Astar :: in_bounds (Node* node) {

	if (node->row() >= 0 and node->row() < this->grid->rows and node->col() >= 0 and node->col() < this->grid->cols) {
		return true;
	} else {
		return false;
	}

}


bool Astar :: is_wall (Node* node) {

	if ((int) this->grid->at<uchar>(node->row(), node->col()) == 0) {
		return true;
	} else {
		return false;
	}
}


vector<Node*> Astar :: get_neighbors (Node* node) {

	vector<Node*> neighbors;
	int i, j;
	for (i = -1; i <= 1; i++) {
		for (j = -1; j <= 1; j++) {
			if (i != 0 or j != 0) {
				Node* neighbor = new Node (node->row() + i, node->col() + j);
				if (this->in_bounds(neighbor) and not this->is_wall(neighbor)) {
					neighbors.push_back(neighbor);
				}
			}

		}
	}

	return neighbors;

}

double Astar :: compute_cost (Node* current, Node* neighbor, Node* start) {

	double v = this->V (neighbor, start);
	double n = this->N (current, neighbor);
	double m = this->M (neighbor);
	double d = this->D (neighbor);
	double d2 = this->D2 (neighbor);

	return 3*v + 1*n + 50*m + 150*d + 50*d2;

}

double Astar :: V (Node* neighbor, Node* start) {

	return abs((neighbor->row () - start->row ()));

}

double Astar :: N (Node* current, Node* neighbor) {

	if (current->row () == neighbor->row () and current->col () == neighbor->col ()) {
		return (double) 10;
	} else {
		return (double) 14;
	}

}

double Astar :: M (Node* node) {

	if (this->is_wall(node)) {
		return (double) 0;
	} else {
		return (double) 1;
	}

}

double Astar :: D (Node* node) {

	return 1 / (1 + min(this->upward_obstacle(node), this->downward_obstacle(node)));

}

double Astar :: D2 (Node* node) {

	double m = min(this->upward_obstacle(node), this->downward_obstacle(node));
	return 1 / (1 + pow(m, 2));

}

double Astar :: upward_obstacle (Node* node) {

	int step = 1;
	while (step <= 50) {
		Node up = Node (node->row() - step, node->col());
		if (is_wall (&up)) {
			return (double) step;
		} else {
			step++;
		}
	}

	return INFINITY;

}

double Astar :: downward_obstacle (Node* node) {

	int step = 1;
	while (step <= 50) {
		Node down = Node (node->row() + step, node->col());
		if (is_wall (&down)) {
			return (double) step;
		} else {
			step++;
		}
	}

	return INFINITY;

}

vector<Node*> Astar :: reconstruct_path (vector<Node*> closedSet) {

	Node* current = closedSet.back();
	vector<Node*> total_path;
	total_path.push_back(current);
	while (current->parent() != NULL) {
		current = current->parent();
		total_path.push_back(current);
	}

	return total_path;

}

vector<Node*> Astar :: pathfind (Node* start, Node* goal) {

	vector<Node*> closedSet;
	priority_queue<Node*, vector<Node*>, CompareFscore> openSet;

	start->gscore(0);
	start->fscore(this->heuristic(start, goal));
	openSet.push(start);

	cout << "start" << endl;

	while (not openSet.empty()) {

		Node* current = openSet.top();
		//cout << *current << endl;
		openSet.pop();
		closedSet.push_back(current);

		if (*current == goal) {
			break;
		}

		vector<Node*> neighbors = this->get_neighbors(current);
		for (unsigned int i = 0; i < neighbors.size(); i++) {
			Node* neighbor = neighbors[i];

			if (find(closedSet.begin(), closedSet.end(), neighbor) != closedSet.end()){
				continue;
			}

			double new_gscore = current->gscore() + this->heuristic(current, neighbor); //this->compute_cost(current, neighbor, start);

			if (neighbor->parent() == NULL or new_gscore < neighbor->gscore()) {
				neighbor->gscore(new_gscore);
				double fscore = new_gscore + this->heuristic(neighbor, goal);
				neighbor->fscore(fscore);
				neighbor->parent(current);
				openSet.push(neighbor);
				//cout << "neighbor: " << *neighbor << " - fscore: " << neighbor->fscore() << " - parent: " << *neighbor->parent() << endl;

			}

		}

	}

	cout << "end" << endl;
	return this->reconstruct_path(closedSet);

}

File: python/linesegm/lib/astar.py
import numpy as np
import heapq


class PriorityQueue():

    def __init__(self):
        self.list = []

    def empty(self):
        return len(self.list) == 0

    def put(self, item, priority):
        heapq.heappush(self.list, (priority, item))

    def get(self):
        return heapq.heappop(self.list)[1]

    def size(self):
        return len(self.list)

    def flush(self):
        size = self.size()
        [self.list.pop() for i in range(size / 2, size - 1)]
        heapq.heapify(self.list)


class Astar():

    def __init__(self, grid):
        self.parents = {}
        self.open = PriorityQueue()
        self.close = []
        self.grid = grid / 255
        self.gscore = np.zeros((self.grid.shape), dtype=np.int32)
        # print self.grid
        self.step = 2

    def pathfind(self, start, goal):
        # initialize
        self.start = start
        self.goal = goal
        self.gscore[tuple(self.start)] = 0
        self.open.put(self.start, self.heuristic(self.start, self.goal))
        print 'start: ' + str(self.start) + " - goal: " + str(self.goal),

        while not self.open.empty():

            current = self.open.get()
            self.print_info(current)

            if current == self.goal:
                return self.reconstruct_path(current)

            self.close.append(current)

            for neighbor in self.get_neighbors(current):

                if neighbor in self.close:
                    continue

                new_gscore = self.gscore[tuple(current)] + self.compute_cost(current, neighbor, self.start)

                if (str(neighbor) not in self.parents or new_gscore < self.get_gscore(tuple(neighbor))):
                    self.gscore[tuple(neighbor)] = new_gscore
                    self.parents[str(neighbor)] = current
                    fscore = new_gscore + self.heuristic(neighbor, self.goal)
                    self.open.put(neighbor, fscore)

        return None

    def heuristic(self, current, goal):
        # return abs(current[0] - goal[0]) + abs(current[1] - goal[1])
        return 40*((current[0] - goal[0])**2 + (current[1] - goal[1])**2)**0.5

    def get_neighbors(self, node):
        r, c = node
        s = self.step
        neighbors = [[r - s, c - s], [r - s, c], [r - s, c + s],
                     [r, c - s], [r, c + s],
                     [r + s, c - s], [r + s, c], [r + s, c + s]]
        return filter(self.in_bounds, neighbors)
        # return filter(self.walkable, neighbors)

    def in_bounds(self, node):
        r, c = node
        return 0 <= r < self.grid.shape[0] and 0 <= c < self.grid.shape[1]

    def walkable(self, node):
        r, c = node
        try:
            if self.in_bounds(node) and self.grid[r, c] == 1:
                return True
            else:
                return False
        except IndexError:
            return False

    def get_gscore(self, node):
        try:
            return self.gscore[tuple(node)]
        except KeyError:
            return float('inf')

    def compute_cost(self, current, neighbor, start):
        v = self.V(neighbor, start)
        n = self.N(current, neighbor)
        m = self.M(neighbor)
        d = self.D(neighbor)
        d2 = self.D2(neighbor)
        return 3*v+1*n+50*m+150*d+50*d2
        # return 2.5*v+1*n+50*m+130*d+0*d2

    def V(self, node, start):
        return abs(node[0] - start[0])

    def N(self, current, neighbor):
        if (current[0] == neighbor[0] or current[1] == neighbor[1]):
            return 10.0
        else:
            return 14.0

    def M(self, node):
        if self.grid[node[0], node[1]] == 1:
            return 0.0
        elif self.grid[node[0], node[1]] == 0:
            return 1.0

    def D(self, neighbor):
        return 1 / (1 + min(self.upward_obstacle(neighbor), self.downward_obstacle(neighbor)))

    def D2(self, neighbor):
        return 1 / (1 + min(self.upward_obstacle(neighbor), self.downward_obstacle(neighbor)) ** 2)

    def upward_obstacle(self, node):
        step = 1
        try:
            while(step <= 50):
                if self.grid[node[0] - step, node[1]] == 0:
                    return float(step)
                else:
                    step += 1

            return float('inf')
        except IndexError:
            return float('inf')

    def downward_obstacle(self, node):
        step = 1
        try:
            while(step <= 50):
                if self.grid[node[0] + step, node[1]] == 0:
                    return float(step)
                else:
                    step += 1

            return float('inf')
        except IndexError:
            return float('inf')

    def reconstruct_path(self, current):
        total_path = [current]
        while str(current) in self.parents:
            current = self.parents[str(current)]
            total_path.append(current)

        return total_path, self.close

    def print_info(self, current):
        # print 'gscore: ' + str(self.gscore)
        print 'current: ' + str(current)
        # print 'open: ' + str(self.open)
        # print 'close: ' + str(self.close)
        # print '------------------------------------'
